Add Rust-based fluent file converter with PO format support

iangmaia opened the pull request:
Related: [AINFRA-723: Develop CLI tool using Rust to properly parse Fluent files](https://linear.app/a8c/issue/AINFRA-723/develop-cli-tool-using-rust-to-properly-parse-fluent-files)

This PR introduces a new Rust-based tool for converting Fluent localization files to other formats, starting with PO (Portable Object) file support.

### What's included:
- **Project setup**: Cargo configuration with necessary dependencies
- **Core functionality**: Shared modules for Fluent parsing and error handling
- **PO converter**: Complete implementation for converting .ftl files to .po format
- **CLI interface**
- **Documentation**: README with usage instructions and examples
- **Tests**: Sample data for validation

### Related PRs
On `fluent-rust-tools`:
- #2
- #3 

On `wordpress-rs`, using the tool via the Ruby Gem:
- https://github.com/Automattic/wordpress-rs/pull/765
- https://github.com/Automattic/wordpress-rs/pull/770

iangmaia committed:
Add project setup and PO file conversion functionality

iangmaia committed:
Add tests for PO converter and parser

iangmaia committed:
Refactor Fluent parser to better parse and support comments and add tests

iangmaia committed:
Fix PO metadata headers issue

iangmaia committed:
Clean-up README

iangmaia committed:
Add option to add original language file when generating a PO (to show source strings)

iangmaia committed:
Update code to better handle empty keys and line breaks

iangmaia commented on README.md:
````diff
@@ -0,0 +1,139 @@
+fluent-tools po to-fluent -i input.po -o output.ftl
+```
+
+### Ruby API
````
**Note**: I broke the tool down into separate PRs to roughly separate the main pieces of the project, but there are probably points where the code depends on things only implemented by the next PRs, like this doc mentioning Android and the Ruby API. Those parts are fully implemented only in https://github.com/Automattic/fluent-rust-tools/pull/2 and https://github.com/Automattic/fluent-rust-tools/pull/3.

AliSoftware commented on Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[dev-dependencies]
+tempfile = "3.8"
+assert_cmd = "2.0"
+predicates = "3.0" 
```
Newline at EOF?

iangmaia replied:
👍 Updated on dcea0abff505035d680d79faf5b9cfda1dec2bf7

AliSoftware commented on README.md:
````diff
@@ -0,0 +1,139 @@
+
+```bash
+# Convert Fluent to Android XML
+fluent-tools android to-xml -i input.ftl -o output.xml
````
💄  For symmetry (both between the 2 directions, and between the various subcommands and various interop formats), maybe we should name those `from-fluent` instead of `to-xml`?

That way  for each command (`android`, `po`, …) the 2 subcommands will always be `from-fluent` and `to-fluent` regardless of the command and intermediate format (making it easy to interchange one for another)

Alternatively, if we want to mirror the Ruby API and also still keep a reference to the intermediate format, maybe `fluent-tools android fluent-to-xml` and `fluent-tools android xml-to-fluent` would work too…

iangmaia replied:
> That way for each command (android, po, …) the 2 subcommands will always be from-fluent and to-fluent regardless of the command and intermediate format (making it easy to interchange one for another)

That's a great idea! I think it will make it clearer / improve the UX. Updated in 82a2be74239aee0d027a86af21b745f40d930a8f

AliSoftware commented on README.md:
```diff
@@ -0,0 +1,139 @@
+
+## Features
+
+### Android XML Support
+- Convert Fluent messages to Android string resources
+- Handle plurals, variables
+- Preserve comments and variable mappings
+- Support for bidirectional conversion
+
+### PO Format Support
+- Convert Fluent to GNU gettext PO format
+- Support for plural forms
+- Preserve message context and comments
+- Support for bidirectional conversion
```
I haven't looks at the PRs #2 and #3 just yet, but maybe since I'm guessing you won't implement support for plural forms right now (at least for .`po`) as well as maybe not yet support preserving comments for Android XML (at least GlotPress doesn't seem to, even if your tool does…), what about introducing checkboxes in that README (like you can see in READMEs of many in-progress OSS dev tools) to differentiate what's already implemented vs what's planned for the future but not yet supported?

iangmaia replied:
In fact, we do support comments and variables preservation and plurals, at least the most common constructs, both for PO and Android.
On Android, I had to introduce the "source" file as argument reference when converting back from Fluent to XML for the variables, iinm. On PO, I added identifiers like `FLUENT_ZERO` to properly match the plurals... anyway, it took some time to get everything right but what was listed here should be ok.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+
+    #[test]
+    fn test_fluent_to_po_simple() {
+        let temp_dir = tempdir().unwrap();
```
I've used ChatGPT a lot to ask it to teach me some various Rust concepts the other weekend as a way to do a quick level-up, and TIL that thanks to Rust's ownership concepts, `tempdir()` is able to auto-delete the temp dir once its owned value goes out of scope, i.e. at the end of the test method 👏

iangmaia replied:
🎯 
That's pretty cool indeed 🙂

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        assert!(po_content.contains("msgid \"Hello World\""));
+        assert!(po_content.contains("msgctxt \"hello\""));
+        assert!(po_content.contains("msgid \"Hello, {$name}!\""));
+        assert!(po_content.contains("msgctxt \"greeting\""));
```
I'd suggest to assert on the whole content of `po_content`, or at least the multiline block containing those for each entry, to ensure that the `msgid` and `msgctxt` that are set in the `po_content` output go by pair.

I'd also expect that each of those blocks in the `.po` would also contain a `mgsstr ""` line we need to verify is present?

iangmaia replied:
👍 Updated in 510505acc8f68a3d9bc3e4aa304fa601cf7ed4b8.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        assert!(output_path.exists());
+        let po_content = fs::read_to_string(&output_path).unwrap();
+        assert!(po_content.contains("msgid_plural"));
+        assert!(po_content.contains("FLUENT_SELECTOR:num"));
```
Similar here; especially I'd make the test ensure that this `FLUENT_SELECTOR:` line is inserted as the expected type of comment in the `.po`

iangmaia replied:
👍 Updated in 510505acc8f68a3d9bc3e4aa304fa601cf7ed4b8.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        catalog.append_or_update(message);
+        
+        // Write the catalog to file
+        write_po_file(&catalog, &input_path).unwrap();
```
I'd argue that the point of the test is to test the conversion on an already generated file, and that the test shouldn't include the responsibility to generate the `.po` file using the `po` crate, but rather use a hardcoded `.po` text file content (similar to how you provided a hardcoded text content for the `fluent_content` on line 58)

iangmaia replied:
Agreed 👍 updated on 41e899e354ac6c6632de58be7124e8e0418bb4b4.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+greeting = Hello, {$name}!
+# This is a comment
+farewell = Goodbye, {$name}!
+"#;
```
💄  ChatGPT suggested we could use the `indoc` crate and macro to allow nicer indentation of multiline strings that would strip leading indentation of lines for you:

```toml
[dev-dependencies]
indoc = "2"
```

```rust
use indoc::indoc;

#[test]
fn test_something() {
    let original_content = indoc! {r#"
        hello = Hello World
        greeting = Hello, {$name}!
        # This is a comment
        farewell = Goodbye, {$name}!
    "#};

    assert!(original_content.contains("greeting = Hello, {$name}!"));
}
```

Obviously this is just a bonus and could be for refinement in later PR, but still figured I'd mention it for easier-to-read tests in the future

iangmaia replied:
I like that 👍 will add it to tackle in a refinement PR.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        // Check that key messages are preserved
+        assert!(converted_content.contains("hello = Hello World"));
+        assert!(converted_content.contains("greeting = Hello, {$name}!"));
+        assert!(converted_content.contains("farewell = Goodbye, {$name}!"));
```
iinm you don't have a dedicated round-trip test for entries with multi-line values, while this seems to be a typical case we'd want to test. Instead of adding a dedicated test for multiline like you did for plurals and comments, you could just make this one use multiline values as its input/output test data.

iangmaia replied:
👍 Updated on 5f9b4b6b625a10776417ff1ec7b4da59cc71a48f.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        // Verify PO file contains comments
+        let po_content = fs::read_to_string(&output_path).unwrap();
+        assert!(po_content.contains("This is a greeting message"));
+        assert!(po_content.contains("Welcome message for users"));
```
Like above, you should assert not just on the string but on the whole line, including `# ` comment prefix or `msgid` prefix, to make the test more resilient.

iangmaia replied:
👍 Updated in 510505acc8f68a3d9bc3e4aa304fa601cf7ed4b8.

AliSoftware commented on src/po/converter.rs:
```diff
@@ -0,0 +1,391 @@
+        assert!(po_content.contains("Welcome message for users"));
+    }
+
+    #[test]
+    fn test_fluent_to_po_invalid_input() {
+        let temp_dir = tempdir().unwrap();
+        let input_path = temp_dir.path().join("nonexistent.ftl");
+        let output_path = temp_dir.path().join("output.po");
+        
+        // Try to convert non-existent file
+        let result = fluent_to_po(&input_path, &output_path, "en", None);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn test_po_to_fluent_invalid_input() {
+        let temp_dir = tempdir().unwrap();
+        let input_path = temp_dir.path().join("nonexistent.po");
+        let output_path = temp_dir.path().join("output.ftl");
+        
+        // Try to convert non-existent file
+        let result = po_to_fluent(&input_path, &output_path);
+        assert!(result.is_err());
+    }
```
Would be nice to have tests for actually invalid input (i.e. an existing `.po` or `.ftl` file but with content containing syntax errors), not just non-existent files like here.

iangmaia replied:
👍 Added on da0cc07796d854248779264a15aa45023ce46b77.

AliSoftware commented on src/shared/error.rs:
```diff
@@ -0,0 +1,52 @@
+    
+    #[allow(dead_code)]
+    #[error("Invalid Android XML format: {0}")]
+    InvalidAndroidXml(String),
```
It feels a bit odd to me that the errors about things like Android XML parsing are in the shared module, as opposed to being in the `android/` module, and same for `po`.

Would feel more appropriate to me to have all the things that are specific to Android XML be in the `android` module, and same for PO Catalog be in the `po` module, to keep each things isolated and separate?

That being said, I don't know enough Rust to know if that's a typical pattern and how to encode that. I'm guessing maybe having generic `InputFileParseError` / `InputFileInvalidFormat` entries (instead of error entries for each format) could be enough? 🤔

iangmaia replied:
> I'm guessing maybe having generic InputFileParseError / InputFileInvalidFormat entries (instead of error entries for each format) could be enough?

Hmm indeed since the modules are isolated and both errors represent a parsing error, I think that's enough.
Updated on 2ad605970609122cf3dbd06d9e092d4e7ba1fb05.

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+        if !attributes.is_empty() {
+            output.push('\n');
+        }
+    }
```
I'm a bit surprised you had to implement this logic yourself?!

The whole point of writing this library in Rust is so that we could use the `fluent_syntax` crate and rely on it to parse and generate `.ftl` files without having to implement parsing/generating ourselves and thus without to reinvent the wheel (and thus without risking missing to account for edge cases etc)

iangmaia replied:
Ah, this was due to a formatting issue I had when converting from Android back to Fluent, and considered it would be ok to have something to iterate the parsed structures and write it out so we could have better control over it. 

I still used the `fluent_syntax` library, just not to generate the final file... but I gather it may look strange and it'd be better to fix the root cause of the Android formatting issue (specially given I had it only on Android, which was the first one I implemented and also as someone not experienced in Rust... so there are probably some rougher edges there).

I've pushed some changes to just use the library on 7bda894baa165d25f57a63a88be5c55f47f01b20.

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+        message: &fluent_syntax::ast::Message<&str>, 
+        message_id: &str
+    ) -> Option<String> {
+        // Priority: 1) message's own comment, 2) pending standalone comment (if close enough)
```
Is this part of the Fluent spec? I thought standalone comments were not supposed to be attached to messages—and thus I wouldn't expect them to be included in the converted file?

iangmaia replied:
Related to https://github.com/Automattic/fluent-rust-tools/pull/1#discussion_r2152908718 (now removed).

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+    if let Some(other_pattern) = variants.get(OTHER_VARIANT) {
+        result.push_str(&format!("   *[{}] {}\n", OTHER_VARIANT, pattern_to_string(other_pattern)));
+    } else {
+        result.push_str("   *[other] (missing)\n");
```
I wonder if we should `panic!` instead here? Or at least return an `Error()`

iangmaia replied:
Updated with the result generation update on https://github.com/Automattic/fluent-rust-tools/commit/7bda894baa165d25f57a63a88be5c55f47f01b20

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+        .join("")
+}
+
+fn extract_element_text(element: &FluentElement) -> String {
+    match element {
+        FluentElement::Text(text) => text.clone(),
```
I wonder if those `extract_…` methods shouldn't be _consuming_, i.e. not take a reference but take ownership of the `FluentElement`—Which would kinda make sense to me semantically because we extract something so we should take care ownership and consume it? And then we wouldn't need the `.clone()` here unlike when you pass a `&` reference…

iangmaia replied:
At the same time, in the current API you "peek" at the content without needing to own the entire structure (which makes sense for a helper function); if we change it consuming, we'd need to clone the entire element on the caller side 🤔

AliSoftware replied:
True…

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+        let hello_comment = resource.messages[0].comment.as_ref().unwrap();
+        assert!(hello_comment.contains("This is a greeting message"));
+        assert!(hello_comment.contains("It supports internationalization"));
+        assert!(hello_comment.contains("and has multiple lines of comments"));
```
Once again I'd suggest to `assert!` on the whole expected comment. For example to ensure that not only does the `hello_comment` contains those strings but it also doesn't contain e.g. unexpected extra indentation or `# ` characters inside the parsed comment itself

iangmaia replied:
👍 Updated in 510505acc8f68a3d9bc3e4aa304fa601cf7ed4b8.

AliSoftware commented on src/shared/fluent_parser.rs:
```diff
@@ -0,0 +1,823 @@
+    #[test]
+    fn test_comment_association_logic() {
+        // Test close association (1 empty line)
+        let ftl_close = r#"
+# This comment IS associated with hello.
+
+hello = Hello
```
Do we need to cover that? (I mean, per the Fluent specs, is such a comment supposed to be associated with the message as opposed as being a standalone comment in the file?)

iangmaia replied:
Now revisiting this part of the code, I recall this is something I specifically implemented but I don't recall the details of the context / reasoning 🙃 perhaps I was running some tests based on specific files.

I think I was worried about missing useful comments with the round trip conversion to PO and then back to Fluent, but now I'd say that's not worth it to worry about (neither conceptually correct). I'm tempted to remove it and just keep regular immediate string comments for now.

AliSoftware replied:
> I'm tempted to remove it and just keep regular immediate string comments for now.

I'd vote for that too

iangmaia replied:
Updated in d615fa2969c643be6ef2f38ea5b250b2b9f65305 which nicely cleaned up the code; from my tests, including a round-trip conversion everything seemed fine, only "loose" comments were lost.

AliSoftware commented on tests/data/sample_source.ftl:
I'm guessing this one is planned to be used in subsequent PRs? As I haven't seen it used in this one at least

iangmaia replied:
Ah, this file was the reference file I used to run a full round trip translation for both Android and PO formats. I do have an integration test in the Makefile (Gem PR) that will use it.

AliSoftware replied:
Ah I vaguely rememing seeing that huge integration test in the `Makefile` now indeed (and my thought back then was I wasn't a fan of having a whole test suite implemented directly in the `Makefile`, as opposed to it being e.g implemented in a `.sh` and `make` just calling it.

AliSoftware reviewed (commented):
It's a bit tough to review with my limited knowledge of Rust (though I used ChatGPT to give me a crash course and I learned a ton in little time!), so I left comments on the things I got.

In terms of code organization this is a bit difficult for me to see the big picture just looking at the "flat" code in a large GitHub diff, and the fact that some of the code in this PR is in anticipation of #2 and #3 and all but that not of it is here yet doesn't help 😅  So I think the best way forward is to start with something functioning like that, iterate with the follow-up PRs you already started, testing it in `wordpress-rs` like you did and then revisit the code in an IDE (instead of flat diff) so I can better grasp the whole picture.

iangmaia committed:
Add line breaks at end of files

iangmaia committed:
Rename all format arguments to be standardized as from-fluent / to-fluent

iangmaia committed:
Update Fluent parser to follow Fluent standard and only link comments on top of strings without empty lines

iangmaia committed:
Improve Fluent / PO test assertions

iangmaia committed:
Use hard-coded PO instead of generating PO via code to test PO to Fluent generation

iangmaia committed:
Update test_round_trip_conversion test to include multiline values

iangmaia committed:
Add improved error handling on PO parsing

iangmaia committed:
Add tests for invalid Fluent / PO syntax content

iangmaia committed:
Remove specific errors in favor of generic ones

iangmaia committed:
Update Fluent output formatting to fully use the lib `fluent_syntax::ast`

iangmaia committed:
Improve to Fluent parsing and writing

iangmaia committed:
Add test to confirm the output of the `fluent_syntax` library

iangmaia commented:
@AliSoftware thanks for the first batch of comments in this PR and #2 / #3 -- I think I've addressed them (or most of them) and the code is in a much better shape (specially regarding some parsing/serialization aspects) 🙏 
There are still a few things I'd like to tweak in future PRs, but I think the baseline and main aspects of the tool seem to be working well, including being [tested](https://github.com/Automattic/wordpress-rs/pull/765/commits/4c71ff277c692eddad8bd3a28dbece5cc0c79665) in `wordpress-rs`.

But I do understand that these PRs are quite big and hard to review and could take a lot of time... so perhaps we could, as long as there isn't anything very obvious to tackle, move fast to integrate the tool into the team's daily process and continue to tweak it in future PRs (as you mentioned [here](https://github.com/Automattic/fluent-rust-tools/pull/1#pullrequestreview-2932305332)).

AliSoftware commented on src/shared/fluent_parser.rs:
WDYT about splitting this file in multiple files instead:
 - One file for `FluentResourceParser` and all its logic for transforming a `String` into the `Fluent*` data structures
 - One file where you'd extract the logic for transforming the `Fluent*` data back to a `String`, i.e. the counterpart of `FluentResourceParser` (maybe call it `FluentResourceWriter`?)
 - Then a file for defining the data objects only (`struct` or `enum` without `impl`, like `FluentMessage`, `FluentPattern`, etc…) as well as the `struct FluentResource` type and its `impl`—which would now only contain `pub fn from_source` and `pub fn to_source` methods and nothing else, because the whole logic that you had in private functions of that `FluentResource` so far would now be extracted in the `FluentResourceWriter` dedicated type

That way:
 - You'd get some symmetry for the implementation when converting between both directions—by having a `FluentResourceWriter` mirroring the `FluentResourceParser`, and abstracting away all the private logic into dedicated classes
 - Then your data structures that are intended to just represent data like a `FluentResource` would be much simpler
 - It'd be easier to read the code by having mostly one class per file (at least for `FluentResourceParser` and `FluentResourceWriter` being now isolated) instead of everything mixed in a giant file like here (which is another reason why I'm been having a hard time reviewing this tbh, with everything mixed in a single long file

iangmaia replied:
Besides the bigger update mentioned on https://github.com/Automattic/fluent-rust-tools/pull/1#discussion_r2169337155, I had also previously implemented this one on e2699e50c6d2591de24f01599a0cb9d442373849.

AliSoftware commented on src/po/po_format.rs:
```diff
@@ -0,0 +1,1052 @@
+    }))
+}
+
+/// Parse a PO plural form string into (variant_key, text)
+fn parse_plural_form(msgstr: &str) -> Option<(String, String)> {
+    if let Some(colon_pos) = msgstr.find(':') {
+        let (marker, text) = msgstr.split_at(colon_pos);
+        let text = text[1..].to_string(); // Remove the ':' character
+        
+        let variant_key = match marker {
+            FLUENT_ZERO_MARKER => "0".to_string(),
+            FLUENT_ONE_MARKER => "one".to_string(),
+            FLUENT_OTHER_MARKER => "other".to_string(),
+            other if other.starts_with(FLUENT_MARKER_PREFIX) => {
+                other[FLUENT_MARKER_PREFIX.len()..].to_lowercase()
+            }
+            _ => "other".to_string(), // Fallback for unknown markers
+        };
+        
+        Some((variant_key, text))
+    } else {
+        // Fallback: treat as 'other' form if no marker
+        Some(("other".to_string(), msgstr.to_string()))
+    }
+}
```
I'm confused… does this implementation implies that the `msgstr` that you're parsing from the original `.po` file here is expected to be `msgstr "one:You liked one post"` and `msgstr "other:You liked {$count} posts"`, with that `one:` and `other:` prefix being as part of the `msgstr` in the PO?! 🤔  As opposed to relying on `msgstr[0]`/`msgstr[1]` that the `.po` format expects for plural forms?

iangmaia replied:
Ah, that's an issue in the PO format. I couldn't find a way of using the regular constructs to represent these possibilities:
```fluent
notification-count = {$count ->
    [0] No new notifications
    [one] You have {$count} new notification
   *[other] You have {$count} new notifications
}

shared-photos = {$photoCount ->
    [one] added a new photo
    *[other] added {$photoCount} new photos
}
```

 I've pushed 346b75ae67c19c9e265482a63d0cf5dd9d686330 to add a bit more context.

AliSoftware replied:
Tbh I wouldn't expect our tool to support anything other than the CLDR named categories (`zero`, `one`, `many`, `others`, etc), not numerical values. So I don't think we should go out of our ways to support cases like `[0]`. We can just support:
```fluent
shared-photos = {$photoCount ->
    [one] added a new photo
    *[other] added {$photoCount} new photos
}
```
```gettext
msgid shared-photos
msgstr[0] added a new photo
msgstr[1] added ${photoCount} new photos
```
(Well, when transforming originals and assuming those originals are `en-US`, you could consider one would only expect the original `ftl` to only contain `[one]` and `[other]` as those are the only plural categories en-US supports, and use `msgstr` and `msgstr_plural` for the original `.pot` too. But for other locales that supports more than just `[one]` and `[other]` plural categories, you'd still use `msgstr[n]`)

This is important to use `msgstr` and `msgstr_plurals` (for original) and `msgstr[n]` for plurals and not add those prefixes to the strings, because those entries in the `.po` will be what GlotPress uses to detect those as a _single_ entry with plurals variations—as opposed to those being interpreted as different 3 entries by GlotPress with 3 different original strings (each prefixed with `0:`/`one:`/`other:`).

We wouldn't want the translators to see those as 3 different singular entries—one to translate a `0:No new notifications` string, another entry to translate `one:You have {$count} new notification` string, and a 3rd to translate `other:You have {$count} new notifications`, and have those be represented as 3 unrelated entries in GlotPress instead of a single unit. That would defeat the whole purpose…

iangmaia replied:
> This is important to use msgstr and msgstr_plural

> We wouldn't want the translators to see those as 3 different singular entries

This wasn't the case, though -- the plural construct was still used (or _abused_) to fit the Fluent cases (`msgstr[0] "FLUENT_ZERO:No new notifications`).

But I'm fine removing it and following the standard, it makes sense... in fact there's an issue that (I think) made me go to this route in the first place: `po_lib` doesn't seem to properly generate the `Plural-Forms` header based on the locale.
I've just tested building a map of those (via Claude) and ran a few tests, I'm tempted to keep it 🤔

AliSoftware replied:
> This wasn't the case, though -- the plural construct was still used (or abused) to fit the Fluent cases (msgstr[0] "FLUENT_ZERO:No new notifications).

Yeah having `msgstr[0] "FLUENT_ZERO:No new notifications` would still be an issue because it would still have the `FLUENT_ZERO:` prefix be included in the string to translate in GlotPress, which would make it super confusing for translators (and also what if they removed or changed that prefix in the translations their provide?)

If you really need to include some metadata about an entry, you'd better use the comment associated with the entry rather than the `msgstr` string, so that at least that data is hidden away in a (non-editable) comment in GlotPress and not part of the string translators will have to translate.

> n fact there's an issue that (I think) made me go to this route in the first place: po_lib doesn't seem to properly generate the Plural-Forms header based on the locale.
> I've just tested building a map of those (via Claude) and ran a few tests, I'm tempted to keep it 🤔

Yep I think having a crude hardcoded map of `locale => Plural-Forms header` is totally fine at that point given the scope of this tool.

iangmaia replied:
Updated the plurals implementation on ceeb04ad7040a8741b5182e317a1115d44a4920f, adding CLDR support with better locale handling.

iangmaia committed:
Add comment about Fluent plural markers

iangmaia committed:
Refactor Fluent parser

iangmaia committed:
Update plurals handling
Remove fluent placeholders generated in the PO, properly use Plural-Forms and msgstr[i]

iangmaia commented:
Hi there @oguzkocer and @crazytonyli -- adding you both to a set of PRs in https://github.com/Automattic/fluent-rust-tools repo as briefly discussed in p1751037947930919-slack-C02KLTL3MKM.

This is more to have more eyes looking into this tool specially given neither myself nor @AliSoftware have experience developing with Rust, and there are probably best practices and other things (like reuse opportunities) we're missing.
If you prefer avoiding long PR feedback loops, as @oguzkocer suggested, I'm totally fine if you create a PR targeting this PR with changes.

### Some background and a quick overview

We came up with a localization workflow where the original (`en`) Fluent file in `wordpress-rs` would be first converted to an exchange format and then uploaded to GlotPress. Translations then would be fetched from GlotPress in that same exchange format and will be converted to Fluent.

We first started with a basic Ruby tooling based on a [Fastlane action](https://github.com/wordpress-mobile/release-toolkit/pull/650) to perform the conversions... but we quickly figured it would be very fragile, and we could instead try and build a parser in Rust using a proper parser and that would be more robust in the long run.

We then created the Rust CLI tool https://github.com/Automattic/fluent-rust-tools, which is embedded in a Ruby Gem and can be easily integrated into our regular Fastlane automation scripts and into CI. I ran a few tests and things seem to be working fine at this point.

### PRs

In `fluent-rust-tools`:
- https://github.com/Automattic/fluent-rust-tools/pull/1: the present PR, with the code supporting bi-directional conversion between Fluent and PO formats.
- https://github.com/Automattic/fluent-rust-tools/pull/2: I consider this PR a "bonus" as we started with the idea of using Android XML as an interchange format for its simplicity compared to PO, but a few issues like GlotPress stripping out comments made us return to the PO approach. We still kept the code as we could perhaps use this tool for similar conversions (or other formats) at some point, perhaps in other projects.
- https://github.com/Automattic/fluent-rust-tools/pull/3: final PR with the CI setup, additional configuration and Ruby Gem setup (Approved ✅). **If you want to run the tool / tests, you need to use this branch** -- I broke down the PRs after having an initial version running, so I just roughly split the code for reviewing.

In `wordpress-rs`:
- https://github.com/Automattic/wordpress-rs/pull/765: PR integrating the RubyGem and adding a couple of lanes to generate .po/Fluent files, download translations from GlotPress (Approved ✅)
- https://github.com/Automattic/wordpress-rs/pull/770: final PR setting up a job to run on a nightly schedule updating the translations. (Approved ✅)

In `wpcom`:
- 184627-ghe-Automattic/wpcom - configuration to sync the original language `.po` file with GlotPress (Approved ✅)

iangmaia committed:
Clean-up README

iangmaia commented:
Following-up to the discussion in p1751037947930919-slack-C02KLTL3MKM, I'll merge this PR and we'll later iterate on the tool as a whole with improvements and eventual fixes once we get a localization process running in `wordpress-rs`.
